### ISSUE TITLE
Rolling update fails sometimes because of a timeout and leaves two replication controllers

### ISSUE THREAD

**Issue by henrymai** [kubernetes/kubernetes#16737]
This is the command that was run along with the output below.
For clarity, the new rc is:
`django-hash-3059e3853a8b8e954cdab63d6b364713f1d8a652`

The old rc is:
`django-hash-daf9e4eeac8ea0b32ed53d5c991474b2c1140db`

```
Found an existing replication controller for app: django
Hashes don't match. Existing: {u'image_hash': u'hash_daf9e4eeac8ea0b32ed53d5c991474b2c1140dba'}, New: {'image_hash': 'hash_3059e3853a8b8e954cdab63d6b364713f1d8a652'}
Performing rolling update
Traceback (most recent call last):
  File "/var/lib/jenkins/jobs/collectr_deploy/workspace/production/kubernetes/backend_stack/deploy_scripts/update_kubernetes_replication_controllers.py", line 70, in <module>
    "--namespace=%s" % args.namespace])
  File "/usr/lib/python2.7/subprocess.py", line 540, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['kubectl', 'rolling-update', u'django-hash-daf9e4eeac8ea0b32ed53d5c991474b2c1140dba', '-f', '/home/ubuntu/kubernetes_3210706b36bf6dbdbacd7fb308ed635231df1da6/django/django-rc.yaml', '--namespace=sandbox']' returned non-zero exit status 1
stdout: Created django-hash-3059e3853a8b8e954cdab63d6b364713f1d8a652
Scaling up django-hash-3059e3853a8b8e954cdab63d6b364713f1d8a652 from 0 to 3, scaling down django-hash-daf9e4eeac8ea0b32ed53d5c991474b2c1140dba from 3 to 0 (keep 3 pods available, don't exceed 4 pods)
Scaling django-hash-3059e3853a8b8e954cdab63d6b364713f1d8a652 up to 1
Scaling django-hash-daf9e4eeac8ea0b32ed53d5c991474b2c1140dba down to 2
Scaling django-hash-3059e3853a8b8e954cdab63d6b364713f1d8a652 up to 2
stderr: error: timed out waiting for any update progress to be made
```

Aftewards, it leaves behind 2 replication controllers:
```
> kubectl get rc --namespace=sandbox
CONTROLLER                                                   CONTAINER(S)   IMAGE(S)                                                                                               SELECTOR                                                                    REPLICAS   AGE
django-hash-3059e3853a8b8e954cdab63d6b364713f1d8a652         django         snipped:5000/django:hash_3059e3853a8b8e954cdab63d6b364713f1d8a652          app=django,image_hash=hash_3059e3853a8b8e954cdab63d6b364713f1d8a652         2          11h
django-hash-daf9e4eeac8ea0b32ed53d5c991474b2c1140dba         django         snipped:5000/django:hash_daf9e4eeac8ea0b32ed53d5c991474b2c1140dba          app=django,image_hash=hash_daf9e4eeac8ea0b32ed53d5c991474b2c1140dba         2          15h
```

Could this be related to: #8676 ?
Looking at the kubectl get events --namespace=sandbox, ~10 hours or so after it occurred didn't show me anything. We didn't get to look at it earlier because we only discovered it in the morning.
I resolved the issue by deleting the old replication controller manually and rescaling the new replication controller manually to 3.

This is my kubectl version:
```
kubectl version
Client Version: version.Info{Major:"1", Minor:"1+", GitVersion:"v1.1.0-ta-1", GitCommit:"bd56609691112cf026f77d017e89b771635bfdd6", GitTreeState:"clean"}
Server Version: version.Info{Major:"1", Minor:"1+", GitVersion:"v1.1.0-ta-1", GitCommit:"bd56609691112cf026f77d017e89b771635bfdd6", GitTreeState:"clean"}
```
The binaries are built off of this repo:
https://github.com/thesamet/kubernetes/tree/v1.1.0-ta-1

**Comment by johanhaleby** [comment]
I'm seeing this as well. To me this happened when the first pod was Running (but in a bad state and got restarted by k8s frequently) but there was an error in the new pod/rc I updated to (the Dockerfil was missing a CMD) and the POD state was `RunContainerError`. The rolling-update thus hanged until I got error message:  

> error: timed out waiting for any update progress to be made

No I can't seem to update to a newer version, I always get this error.

**Comment by nikhiljindal** [comment]
Yes, rolling-update will not proceed until the new pod is healthy.
You can rerun the same `kubectl rolling-update` command and it will pick up the state from where it left of.

cc @ironcladlou @janetkuo

**Comment by ernestoalejo** [comment]
@nikhiljindal If the pod is failing probably it should be changed before a new rolling-update (like a new  image name without the error). A rollback is needed before starting a new update but it returns a timeout too. The only option left is manually retrieving the new RC name and running a `kubectl delete rc XX --cascade`.

**Comment by nikhiljindal** [comment]
To revert to the previous state (state before rolling update was started), you can run the same `kubectl rolling-update` command with `--rollback` flag: https://github.com/kubernetes/kubernetes/blob/master/docs/design/simple-rolling-update.md#aborting-a-rollout

**Comment by ernestoalejo** [comment]
It timeouts too, it doesn't finish the rollback; even if it has to remove a single failing pod and the new RC.

**Comment by johanhaleby** [comment]
What I had to do was to stop both the old and new replication controller (rc) and recreate the rc with a working set of pods (`kubectl create -f ...`). 

What I'm missing in k8s is a "createOrUpdate" command that either just creates the resources if they don't exist, does a rolling update if an another version exists or stops the rc of the previous version and recreates the rc (and pods) if deemed necessary (because of a failure). Perhaps there are cases where it's unknown what options is preferred and then I'm fine that a user interaction is necessary. We're think of implementation something like this ourselves for our CI server. We're pushing each commit to our test cluster and failures will happen there quite a lot. We don't want to manually have to do rollback etc if there was a failure in the previous commit.

**Comment by ernestoalejo** [comment]
Related issue: #18568

**Comment by ernestoalejo** [comment]
Also related: #18530

**Comment by ernestoalejo** [comment]
I am wondering if deployments will make this issue obsolete. The proposal [1] says old RCs will be deleted; so we should not have any failed RC once the new successful deployment takes over.

[1] https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/deployment.md

**Comment by jsravn** [comment]
We do a rollback on failed rolling update, but it is a separate step. I thought it was done this way on purpose to give people flexibility to decide how to handle a failed update. But I'd be happy having it be a flag that does the rollback automatically.

A couple other issues are stopping the rollback from working though, the ones @ernestoalejo listed.

Deployment api looks good but I don't want to rely on an alpha/experimental API right now.

**Comment by nikhiljindal** [comment]
Yes with deployments API, you can set maxUnavailable to 100% and then it will always update old pods with new ones, irrespective of whether they are failing or not.

We plan to graduate deployments to beta pretty soon.

Also, you guys should check out the proposal for rolling-update like command using deployments API: https://github.com/kubernetes/kubernetes/pull/18404

**Comment by azr** [comment]
I'm having the issue except nothing fails, it's just a pod takes at least 1h to be ready.
To change this set you `--timeout=` to something really bigger than your boot time.

**Comment by knownasilya** [comment]
I'm only seeing this with kubectl 1.1.3. When I rollback to the previous version (`gcloud components restore`) it works just fine.

**Comment by bgrant0607** [comment]
@ernestoalejo pointed out a serious issue with rollback: It needs to be possible to rollback when pods are unavailable, since that's likely when one wants to rollback. The user should be able to do this without changing maxUnavailable or maxSurge.

The deployment controller currently just checks whether we've hit maxUnavailable in order to decide whether it can scale down:
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/deployment/deployment_controller.go#L643
```
	if readyPodCount <= minAvailable {
		// Cannot scale down.
		return false, nil
	}
```

However, this overlooks a critical feature of ReplicationController/ReplicaSet: It favors killing pods that aren't serving:
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/replication/replication_controller.go#L377
```
			// Sort the pods in the order such that not-ready < ready, unscheduled
			// < scheduled, and pending < running. This ensures that we delete pods
			// in the earlier stages whenever possible.
			sort.Sort(controller.ActivePods(filteredPods))
```

Therefore, it should be safe to scale down unhealthy replicas, since that won't further increase unavailability.

This should be permitted generally -- we shouldn't need to special-case rollback.

cc @mqliang

**Comment by bgrant0607** [comment]
@johanhaleby 

kubectl apply + Deployment will support "createOrUpdate":
https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/deployment.md
https://github.com/kubernetes/kubernetes/blob/master/docs/user-guide/kubectl/kubectl_apply.md

We're also working on automatically reverting a rollback that doesn't make progress: #19343

The intent of Deployment is for it to facilitate CI/CD. It was inspired by Openshift's DeploymentConfig API, which is used for that purpose (#1743).

**Comment by bgrant0607** [comment]
@knownasilya kubectl rolling-update now takes into account pod readiness when deciding whether it should continue, so it doesn't blindly kill all the replicas of the application.

**Comment by bgrant0607** [comment]
The rollback problem was previously discovered in kubectl rolling-update as well: 
https://github.com/kubernetes/kubernetes/issues/18530#issuecomment-163990803

**Comment by bgrant0607** [comment]
@mqliang Is this something you'd be interested in working on in Deployment? I'd like to get it into 1.2.

**Comment by mqliang** [comment]
@bgrant0607 I am interested. I will work on it.

**Comment by bgrant0607** [comment]
@mqliang Great, thanks.

FYI, Deployment PRs in flight:
https://github.com/kubernetes/kubernetes/pulls/janetkuo
https://github.com/kubernetes/kubernetes/pulls/kargakis
https://github.com/kubernetes/kubernetes/pulls/madhusudancs
#18077

**Comment by bgrant0607** [comment]
This is at least one of the places that needs to be changed:
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/deployment/deployment_controller.go#L628
```
	// Check if we can scale down.
	minAvailable := deployment.Spec.Replicas - maxUnavailable
```

**Comment by awan1** [comment]
I'm seeing this issue with doing a [simple rolling update](http://stackoverflow.com/a/30194982/2452770), where I'm just changing the image being used. I wasn't able to figure out if this is a problem being resolved in this issue, or if the work done to close this issue will require me to use Deployments rather than doing a CLI `rolling-update`. Could someone clarify?

**Comment by awan1** [comment]
Also, is there a way to enable the Deployments API when [creating a cluster container](https://cloud.google.com/container-engine/docs/clusters/operations#creating_a_container_cluster) on GKE? I couldn't figure out how to apply the [enabling Deployments](http://kubernetes.io/v1.1/docs/user-guide/deployments.html#enabling-deployments-on-kubernetes-cluster) instructions to GKE clusters.

**Comment by awan1** [comment]
Tracked down the rolling update problem to the Pod failing to schedule (PodExceedsFreeCPU, as seen in `kubernetes get events`) and thus the rolling-update hangs and timeouts. 

Probably a separate issue: if PodExceedsFreeCPU and if I have autoscaling enabled for this cluster with a CPU usage rule in effect, why doesn't the cluster increase the number of instances to handle this?

**Comment by bgrant0607** [comment]
@awan1 Deployment can't be used on GKE yet, no. Not until Kubernetes 1.2 is released.

**Comment by bgrant0607** [comment]
This was fixed in Deployment by #20368.

Leaving this open to at least document the problem for kubectl rolling-update.

**Comment by bgrant0607** [comment]
#21273 has been merged.
